WordPressLint/src/main/java/org/wordpress/android/lint/MissingNullAnnotationDetector.kt

```diff
@@ -0,0 +1,157 @@
+import com.intellij.psi.PsiPrimitiveType
+import org.jetbrains.uast.UAnnotated
+import org.jetbrains.uast.UAnnotationMethod
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.UEnumConstant
+import org.jetbrains.uast.UField
+import org.jetbrains.uast.UMethod
+import org.jetbrains.uast.UVariable
+
+class MissingNullAnnotationDetector: Detector(), SourceCodeScanner {
+    override fun getApplicableUastTypes(): List<Class<out UElement>> = listOf(
+            UField::class.java,
+            UMethod::class.java,
+    )
+
+    override fun createUastHandler(context: JavaContext): UElementHandler? = with(context) {
+        if (!isJava(uastFile?.sourcePsi)) {
+            return null
+        }
+        object: UElementHandler() {
+            override fun visitField(node: UField) {
+                if (node.requiresNullAnnotation && !node.isNullAnnotated) {
+                    report(node, MISSING_FIELD_ANNOTATION)
+                }
+            }
+
+            override fun visitMethod(node: UMethod) {
+                if (node.requiresNullAnnotation && !node.isNullAnnotated) {
+                    report(node, MISSING_METHOD_RETURN_TYPE_ANNOTATION)
+                }
+
+                node.uastParameters.forEach { parameter ->
+                    if (parameter.requiresNullAnnotation && !parameter.isNullAnnotated) {
+                        report(parameter, MISSING_METHOD_PARAMETER_ANNOTATION)
+                    }
+                }
+            }
+        }
+    }
+    companion object {
+        val MISSING_FIELD_ANNOTATION = Issue.create(
+                id = "MissingNullAnnotationOnField",
+                briefDescription = "Nullable/NonNull annotation missing on field",
+                explanation = "Checks for missing `@NonNull/@Nullable` annotations for fields.",
+        )
+        val MISSING_METHOD_RETURN_TYPE_ANNOTATION = Issue.create(
+                id = "MissingNullAnnotationOnMethodReturnType",
+                briefDescription = "Nullable/NonNull annotation missing on method return type",
+                explanation = "Checks for missing `@NonNull/@Nullable` annotations on method return types.",
+        )
+        val MISSING_METHOD_PARAMETER_ANNOTATION = Issue.create(
+                id = "MissingNullAnnotationOnMethodParameter",
+                briefDescription = "Nullable/NonNull annotation missing on method parameter",
+                explanation = "Checks for missing `@NonNull/@Nullable` annotations on method parameters.",
+        )
+
+        @Suppress("ForbiddenComment")
+        // TODO: Create a LintFix to normalize our usage to Android's annotations. This can be
+        // useful for running Nullability analysis. See:
+        // https://developer.android.com/studio/write/annotations#nullability-analysis
+        val acceptableNullAnnotations = listOf(
+                "androidx.annotation.NonNull",
+                "androidx.annotation.Nullable",
```

Comment:
I'm not sure whether we want to expand this, or use this as an opportunity to normalize our use of these annotations. Though many annotations are compatible with Kotlin's type system, the note linked in the comment suggests that the `androidx` annotations may have some advantages in Android Studio. I'm leaning toward enforcing the usage of these variants, since this is still only informational anyway. 🤔

Reply:
Yea, I agree on that, I also suggest us using the `androidx` annotations only and replace any usages of non `androidx` annotations in any of our projects.

Having said that, I am not sure about the purpose of this `TODO` comment and what kind of `LintFix` we want to add here, or whether this might be considered out-of-scope. As such, I would personally vote in favor of removing this `TODO` comment and maybe creating a separate issue for that matter. This newly created issue on this repo will then deal with the `androidx` vs. non `androidx` annotation usage. Wdyt? 🤔

Reply:
> I am not sure about the purpose of this TODO comment and what kind of LintFix we want to add here, or whether this might be considered out-of-scope.

The reason I left this comment is that as it stands, the rule will inform about missing annotations _even when a valid annotation is present_. This has potential to be confusing, since our goal with this rule was originally aimed at null analysis in the context of Java <-> Kotlin interop (the largest source of our NPEs). So, it was partially a "note to self", and as such, having it as a separate issue would serve the same purpose - but this was simply an easier way to raise the point for discussion. That said, I think we should consider normalizing our usages of null annotations throughout our projects - out of scope here, but related. Until that is done, these informational warnings could be confusing.

Reply:
Resolved in: https://github.com/wordpress-mobile/WordPress-Lint-Android/pull/11/commits/e29edfe1863cdf6ebbf65c3b98155de8fcef9f14

Reply:
Awesome, thanks for this update @mkevins ! 💯

> That said, I think we should consider normalizing our usages of null annotations throughout our projects

FYI: I just created these 2 issues below, marked them as `in-progress`, and will be working on them today:

1. JP/WPAndroid -> [Nullability Annotations to Java Classes - Replace `org.jetbrains` with `androidx` Annotations #18924](https://github.com/wordpress-mobile/WordPress-Android/issues/18924)
2. FluxC -> [Nullability Annotations to Java Classes - Replace `org.jetbrains` with `androidx` Annotations #2803](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2803)

Reply:
Thanks Petros!